File: src/LinuxApi/SpiDevice.cs
using System;
using System.IO;
using System.Runtime.InteropServices;

namespace LinuxApi
{
    /// <summary>
    /// Driver class for SPI transfers. Most of this is copied from the System.Device.Gpio library
    /// </summary>
    public sealed class SpiDevice : IDisposable
    {
        private const uint SPI_IOC_MESSAGE_1 = 0x40206b00;
        private int _deviceFileDescriptor = -1;
        private readonly uint _speed;

        /// <summary>
        /// Initialize an SPI device
        /// </summary>
        /// <param name="devNode">Path to the /dev node</param>
        /// <param name="speed">Transfer speed in Hz</param>
        public unsafe SpiDevice(string devNode, int speed, int transferMode)
        {
            _speed = (uint)speed;

            _deviceFileDescriptor = Interop.open(devNode, FileOpenFlags.O_RDWR);
            if (_deviceFileDescriptor < 0)
            {
                throw new IOException($"Error {Marshal.GetLastWin32Error()}. Can not open SPI device file '{devNode}'.");
            }

            UnixSpiMode mode;
            switch (transferMode)
            {
                case 0:
                    mode = UnixSpiMode.SPI_MODE_0;
                    break;
                case 1:
                    mode = UnixSpiMode.SPI_MODE_1;
                    break;
                case 2:
                    mode = UnixSpiMode.SPI_MODE_2;
                    break;
                case 3:
                    mode = UnixSpiMode.SPI_MODE_3;
                    break;
                default:
                    throw new ArgumentException($"Transfer mode '{transferMode}' not regignized. Must be between 0 and 3.");
            }
            IntPtr nativePtr = new IntPtr(&mode);

            int result = Interop.ioctl(_deviceFileDescriptor, (uint)SpiSettings.SPI_IOC_WR_MODE, nativePtr);
            if (result == -1)
            {
                throw new IOException($"Error {Marshal.GetLastWin32Error()}. Can not set SPI mode to {mode}.");
            }

            byte dataLengthInBits = 8;
            nativePtr = new IntPtr(&dataLengthInBits);

            result = Interop.ioctl(_deviceFileDescriptor, (uint)SpiSettings.SPI_IOC_WR_BITS_PER_WORD, nativePtr);
            if (result == -1)
            {
                throw new IOException($"Error {Marshal.GetLastWin32Error()}. Can not set SPI data bit length to 8.");
            }

            nativePtr = new IntPtr(&speed);

            result = Interop.ioctl(_deviceFileDescriptor, (uint)SpiSettings.SPI_IOC_WR_MAX_SPEED_HZ, nativePtr);
            if (result == -1)
            {
                throw new IOException($"Error {Marshal.GetLastWin32Error()}. Can not set SPI clock frequency to {speed}.");
            }
        }

        /// <summary>
        /// Finalizer of this class
        /// </summary>
        ~SpiDevice() => Dispose(false);

        /// <summary>
        /// Disposes this instance
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Indicates if this instance has been disposed
        /// </summary>
        private bool disposed = false;

        /// <summary>
        /// Dispose this instance internally
        /// </summary>
        /// <param name="disposing">Release managed resourcess</param>
        private void Dispose(bool disposing)
        {
            if (disposed)
            {
                return;
            }

            if (_deviceFileDescriptor >= 0)
            {
                Interop.close(_deviceFileDescriptor);
                _deviceFileDescriptor = -1;
            }

            disposed = true;
        }

        /// <summary>
        /// Writes and reads data from the SPI device.
        /// </summary>
        /// <param name="writeBuffer">The buffer that contains the data to be written to the SPI device.</param>
        /// <param name="readBuffer">The buffer to read the data from the SPI device.</param>
        public unsafe void TransferFullDuplex(ReadOnlySpan<byte> writeBuffer, Span<byte> readBuffer)
        {
            if (writeBuffer.Length != readBuffer.Length)
            {
                throw new ArgumentException($"Parameters '{nameof(writeBuffer)}' and '{nameof(readBuffer)}' must have the same length.");
            }

            fixed (byte* writeBufferPtr = writeBuffer)
            {
                fixed (byte* readBufferPtr = readBuffer)
                {
                    Transfer(writeBufferPtr, readBufferPtr, writeBuffer.Length);
                }
            }
        }

        private unsafe void Transfer(byte* writeBufferPtr, byte* readBufferPtr, int buffersLength)
        {
            var tr = new spi_ioc_transfer()
            {
                tx_buf = (ulong)writeBufferPtr,
                rx_buf = (ulong)readBufferPtr,
                len = (uint)buffersLength,
                speed_hz = _speed,
                bits_per_word = 8,
                delay_usecs = 0
            };

            int result = Interop.ioctl(_deviceFileDescriptor, SPI_IOC_MESSAGE_1, new IntPtr(&tr));
            if (result < 1)
            {
                throw new IOException($"Error {Marshal.GetLastWin32Error()} performing SPI data transfer.");
            }
        }
    }
}

File: src/DuetRealtimeCore/src/Hardware/SpiDevice.cpp
#include "SpiDevice.h"

#include <fcntl.h>
#include <linux/spi/spidev.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include <cerrno>
#include <cstring>
#include <system_error>

namespace Duet::Sbc
{

	SpiDevice::SpiDevice(const std::string& devNode, uint32_t speedHz, int transferMode)
		: m_fd(::open(devNode.c_str(), O_RDWR))
		, m_speed(speedHz)
	{

		if (m_fd < 0)
		{
			throw std::system_error(errno, std::generic_category(), "Cannot open SPI device '" + devNode + "'");
		}

		try
		{
			uint8_t mode = 0;
			switch (transferMode)
			{
			case 0:
				mode = SPI_MODE_0;
				break;
			case 1:
				mode = SPI_MODE_1;
				break;
			case 2:
				mode = SPI_MODE_2;
				break;
			case 3:
				mode = SPI_MODE_3;
				break;
			default:
				throw std::system_error(EINVAL, std::generic_category(), "SPI transfer mode must be between 0 and 3");
			}
			if (::ioctl(m_fd, SPI_IOC_WR_MODE, &mode) < 0)
			{
				throw std::system_error(errno, std::generic_category(), "Cannot set SPI mode");
			}

			uint8_t bitsPerWord = 8;
			if (::ioctl(m_fd, SPI_IOC_WR_BITS_PER_WORD, &bitsPerWord) < 0)
			{
				throw std::system_error(errno, std::generic_category(), "Cannot set SPI word size to 8 bits");
			}

			if (::ioctl(m_fd, SPI_IOC_WR_MAX_SPEED_HZ, &m_speed) < 0)
			{
				throw std::system_error(errno, std::generic_category(), "Cannot set SPI speed");
			}
		}
		catch (...)
		{
			::close(m_fd);
			m_fd = -1;
			throw;
		}
	}

	SpiDevice::~SpiDevice()
	{
		if (m_fd >= 0)
		{
			::close(m_fd);
			m_fd = -1;
		}
	}

	void SpiDevice::TransferFullDuplex(const uint8_t* tx, const uint8_t* rx, size_t length) const
	{
		spi_ioc_transfer transfer{};
		std::memset(&transfer, 0, sizeof(transfer));
		transfer.tx_buf = reinterpret_cast<uint64_t>(tx);
		transfer.rx_buf = reinterpret_cast<uint64_t>(rx);
		transfer.len = static_cast<uint32_t>(length);
		transfer.speed_hz = m_speed;
		transfer.bits_per_word = 8;

		if (::ioctl(m_fd, SPI_IOC_MESSAGE(1), &transfer) < 1)
		{
			throw std::system_error(errno, std::generic_category(), "SPI transfer failed");
		}
	}

} // namespace Duet::Sbc
